sdks/python/apache_beam/ml/inference/tensorrt_inference.py

```diff
@@ -0,0 +1,213 @@
+  However, since inference_args is an argument in the RunInference interface,
+  we want to make sure it is not passed here in TensorRT's implementation of
+  RunInference.
+  """
+  if inference_args:
+    raise ValueError(
+        'inference_args were provided, but should be None because TensorRT '
+        'engines do not need extra arguments in their execute_v2() call.')
+
+
+class TensorRTEngineHandlerNumPy(ModelHandler[np.ndarray,
+                                              PredictionResult,
+                                              trt.ICudaEngine]):
+  def __init__(self, min_batch_size: int, max_batch_size: int, **kwargs):
+    """Implementation of the ModelHandler interface for TensorRT.
+
+    Example Usage:
+      pcoll | RunInference(
+        TensorRTEngineHandlerNumPy(
+          min_batch_size=1,
+          max_batch_size=1,
+          engine_path="my_uri"))
+
+    Args:
+      min_batch_size: minimum accepted batch size.
+      max_batch_size: maximum accepted batch size.
+      kwargs: Additional arguments like 'engine_path' and 'onnx_path' are
+      currently supported.
+
+    See https://docs.nvidia.com/deeplearning/tensorrt/api/python_api/
+    for details
+    """
+    self.min_batch_size = min_batch_size
+    self.max_batch_size = max_batch_size
+    if 'engine_path' in kwargs:
+      self.engine_path = kwargs.get('engine_path')
+    elif 'onnx_path' in kwargs:
+      self.onnx_path = kwargs.get('onnx_path')
+
+    trt.init_libnvinfer_plugins(LOGGER, namespace="")
+
+  def batch_elements_kwargs(self):
+    """Sets min_batch_size and max_batch_size of a TensorRT engine."""
+    return {
+        'min_batch_size': self.min_batch_size,
+        'max_batch_size': self.max_batch_size
+    }
+
+  def load_model(self) -> trt.ICudaEngine:
+    """Loads and initializes a TensorRT engine for processing."""
+    return _load_engine(self.engine_path)
+
+  def load_onnx(self) -> trt.INetworkDefinition:
+    """Loads and parses an onnx model for processing."""
+    return _load_onnx(self.onnx_path)
+
+  #
+  def build_engine(self, network: trt.INetworkDefinition) -> trt.ICudaEngine:
+    """Build an engine according to parsed/created network."""
+    return _build_engine(network)
+
+  def run_inference(
+      self,
+      batch: np.ndarray,
+      engine: trt.ICudaEngine,
+      inference_args: Optional[Dict[str, Any]] = None
+  ) -> Iterable[PredictionResult]:
+    """
+    Runs inferences on a batch of Tensors and returns an Iterable of
+    TensorRT Predictions.
+
+    Args:
+      batch: A np.ndarray or a np.ndarray that represents a concatenation
+        of multiple arrays as a batch.
+      engine: A TensorRT engine.
+      inference_args: Any additional arguments for an inference
+        that are not applicable to TensorRT.
+
+    Returns:
+      An Iterable of type PredictionResult.
+    """
+    _validate_inference_args(inference_args)
+    context = engine.create_execution_context()
```

Comment:
Context creation should be moved out of `run_inference`

Reply:
yes, just for some context, anything that should only be done once per thread should be done in load_model if possible.

Reply:
@pranavm-nvidia pointed out a great point. My original design was context creation in load_engine(), the way it should be done. However, this will be one more additional argument I will have to pass to run_inference(), hence it will break compatibility with RunInference() transform located in base.py. As a result pipeline tests will not work. Passing context as inference_args is not a good option because it goes contrary to the idea of inference_args (additional optional argument for inference call in pytorch). @ryanthompson591 @yeandy please propose your solution how to resolve this.

Reply:
@ryanthompson591 @yeandy Give me couple hours, I think if I take out memory allocation out of run_inference(), I can pass context instead of engine and we won't have to change anything in RunInference() transform.

Reply:
Thanks for looking into the various solutions.

1. We can denote this subpar version as v1 if we can't find a better workaround. (See proposed solution below)
2. It wouldn't be ideal to break the RunInference API. It would be confusing to users to have an API whose pattern is different than the other RunInference transforms.
3.  This is a possibility as we iterate on RunInference with new features. Probably won't happen in the short term though as we'd have to revisit our designs.

___
**Another option?**

As you know, the `run_inference()` API requires there to be a `model` passed. (see [base.py](https://github.com/apache/beam/blob/master/sdks/python/apache_beam/ml/inference/base.py#L94
)). I think you're currently passing in `engine: trt.ICudaEngine` (which by the way would probably break our type checking? though I haven't looked into it in too much detail yet) as the "model". 

One thing we could do is instead of making the parameter `engine: trt.ICudaEngine`, we could define a custom "model" class that contains the necessary parameters. So instead of this
```
def load_model(self) -> trt.ICudaEngine:
  """Loads and initializes a TensorRT engine for processing."""
  return _load_engine(self.engine_path)
```

We could do
```
class TensorRTModel:
  def __init__(self, engine, context, allocations):
     self.engine = engine
     self.context = context
     self.allocations = allocations
     ...
     
  # Feel free to rename this function
  def get_model_components():
    return self.engine, self.context, self.allocations, ..., ...

def load_model(self) -> :
  """Loads and initializes a TensorRT engine for processing."""
  engine = _load_engine(self.engine_path)
  # more logic to get allocations, etc.
  return TensorRTModel(engine, engine.create_execution_context(), ...)
  
def run_inference(
    self,
    batch: np.ndarray,
    model: TensorRTModel,
    inference_args: Optional[Dict[str, Any]] = None
) -> Iterable[PredictionResult]:
  _validate_inference_args(inference_args)
  engine, context, allocations, ..., ... = model.get_all_components()
  
```

Reply:
@yeandy was also thinking about it couple hours ago. Seems like a good idea. Let me think of how to implement it in practice. How much time do we have to make changes? July 19th? Or you want everything finalized for version 1 before that?

Reply:
I actually initially tried to implement TensorRTEngineHandlerNumPy() with parameters like engine, builder, context, etc. However when used with Pipelines some part of Apache Beam would complain that objects are not picklable. Obviously, trt.Builder, trt.Runtime, trt.ICudaEngine are not picklable, hence I had to resort to TensorRTEngineHandlerNumPy() handling only simple parameters like strings and integers. 
Let me see if I can overcome this issue with what you've suggested.

Reply:
To add some context that might help.

In beam you have the runner and the workers. If you were to start beam up it would make a graph (in the runner) then pickle all the classes/methods/data and send that to the workers.

runner (makes graph) -> pickled -> workers (unpickle graph)

So my question would be, at which point does TensorRTEngineHandlerNumPy get instantiated? My feeling is that if it's in load model, it should work, but if it's in your runner code it might fail (because it fails to pickle).

Reply:
Great to hear, thanks! And agreed; give it a go, but multiple inputs can come later :)

Reply:
Resolved